src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskActivityDisplayDriver.cs

```diff
@@ -1,35 +1,82 @@
     where TActivity : NotifyUserTaskActivity
     where TEditViewModel : NotifyUserTaskActivityViewModel, new()
 {
+    private readonly IHtmlSanitizerService _htmlSanitizerService;
+    private readonly ILiquidTemplateManager _liquidTemplateManager;
+    private readonly NotificationOptions _notificationOptions;
+
+    protected readonly IStringLocalizer S;
+
     protected virtual string EditShapeType => $"{nameof(NotifyUserTaskActivity)}_Fields_Edit";
 
+    public NotifyUserTaskActivityDisplayDriver(
+        IHtmlSanitizerService htmlSanitizerService,
+        ILiquidTemplateManager liquidTemplateManager,
+        IOptions<NotificationOptions> notificationOptions,
+        IStringLocalizer stringLocalizer)
+    {
+        _htmlSanitizerService = htmlSanitizerService;
+        _liquidTemplateManager = liquidTemplateManager;
+        _notificationOptions = notificationOptions.Value;
+        S = stringLocalizer;
+    }
+
     public override IDisplayResult Edit(TActivity model)
     {
-        return Initialize(EditShapeType, (Func<TEditViewModel, ValueTask>)(viewModel =>
+        return Initialize<TEditViewModel>(EditShapeType, viewModel =>
         {
             return EditActivityAsync(model, viewModel);
-        })).Location("Content");
+        }).Location("Content");
     }
 
     public async override Task<IDisplayResult> UpdateAsync(TActivity model, IUpdateModel updater)
     {
         var viewModel = new TEditViewModel();
         if (await updater.TryUpdateModelAsync(viewModel, Prefix))
         {
-            await UpdateActivityAsync(viewModel, model);
+            if (!_liquidTemplateManager.Validate(viewModel.Subject, out var subjectErrors))
+            {
+                updater.ModelState.AddModelError(Prefix, nameof(viewModel.Subject), S["Subject field does not contain a valid Liquid expression. Details: {0}", string.Join(' ', subjectErrors)]);
+            }
+
+            if (!_liquidTemplateManager.Validate(viewModel.Summary, out var summaryErrors))
+            {
+                updater.ModelState.AddModelError(Prefix, nameof(viewModel.Summary), S["Summary field does not contain a valid Liquid expression. Details: {0}", string.Join(' ', summaryErrors)]);
+            }
+
+            if (!_liquidTemplateManager.Validate(viewModel.TextBody, out var textBodyErrors))
+            {
+                updater.ModelState.AddModelError(Prefix, nameof(viewModel.TextBody), S["Text Body field does not contain a valid Liquid expression. Details: {0}", string.Join(' ', textBodyErrors)]);
+            }
+
+            if (!_liquidTemplateManager.Validate(viewModel.HtmlBody, out var htmlBodyErrors))
+            {
+                updater.ModelState.AddModelError(Prefix, nameof(viewModel.HtmlBody), S["HTML Body field does not contain a valid Liquid expression. Details: {0}", string.Join(' ', htmlBodyErrors)]);
+            }
+
+            if (updater.ModelState.IsValid)
```

Comment:
We should not do this at all in the UpdateAsync. Just update the model and send it back to the EditAsync with an UpdateEditorContext.

Reply:
@Skrypt This is done in an abstract class which call a virtual method `UpdateActivityAsync` only if the model is valid. We do similar logic in other base drivers like `ActivityDisplayDriver`

Why should we not do this?

Reply:
Going to sleep will try to explain tomorrow. See my display driver pr.

Reply:
I think I will just wait for next triage meeting and explain this. Or I should do a video that explains it all.

Reply:
Sounds good. I suggest opening an issue and explain it with video. Maybe a good topic for Tuesday as well. If this should not be done here, then we should change it in other places too since it is a behavior that we use in other places.

Reply:
I was hoping to follow up on this today on meeting but it will be more likely for next time..

Reply:
No meeting today

Reply:
I saw on Gitter 👍🏼